### PR TITLE
support for creating extensions on a different database than provider

### PR DESCRIPTION
**Pull request by dvdliao**

### CODE REVIEW

**Comment by cyrilgdn** [comment]
@dvdliao Thanks for your work :+1: I'll take a look!

I feel a bit ashamed because I already developed the similar code for my company but didn't take the time to create the PR yet.....

File: postgresql/resource_postgresql_extension.go [hunk 1]
```diff
@@ -101,8 +121,16 @@ func resourcePostgreSQLExtensionExists(d *schema.ResourceData, meta interface{})
 	defer c.catalogLock.Unlock()
 
 	var extensionName string
+
+	database := getDatabaseForExtension(d, c)
+	txn, err := startTransaction(c, database)
```
Comment (cyrilgdn) [inline postgresql/resource_postgresql_extension.go]: You need to check that db exists before trying to connect on it:

```go
       // Check the database exists
       exists, err := dbExists(c.DB(), database)
       if err != nil || !exists {
               return false, err
       }
```
[`dbExists`](https://github.com/terraform-providers/terraform-provider-postgresql/blob/62bb8a15977f56dbebcb5570a41bbe3e5146f0e7/postgresql/helpers.go#L156) can be modified to support both `sql.Tx` and `sql.DB` by creating an interface (already had to do that and inspired from an issue in `database/sql`):

```go
// QueryAble is common interface for sql.DB and sql.Tx
// (cf https://github.com/golang/go/issues/14468)
type QueryAble interface {
	Exec(query string, args ...interface{}) (sql.Result, error)
	Query(query string, args ...interface{}) (*sql.Rows, error)
	QueryRow(query string, args ...interface{}) *sql.Row
}

[...]
func dbExists(db QueryAble, dbname string) (bool, error) {
[...]
```

File: postgresql/resource_postgresql_extension.go [hunk 2]
```diff
@@ -244,9 +300,29 @@ func setExtVersion(db *sql.DB, d *schema.ResourceData) error {
...
-	if _, err := db.Exec(sql); err != nil {
+	if _, err := txn.Exec(sql); err != nil {
 		return errwrap.Wrapf("Error updating extension version: {{err}}", err)
 	}
 
 	return nil
 }
+
+func getDatabaseForExtension(d *schema.ResourceData, client *Client) string {
+	database := client.databaseName
```
[the remark this answers is not in the document]
Comment (dvdliao) [reply postgresql/resource_postgresql_extension.go]: yup, just tested and the old state file id was updated to new format

**Comment by dvdliao** [comment]
Hey @cyrilgdn thanks for the review, I've addressed the comments

File: postgresql/resource_postgresql_extension.go [hunk 3]
```diff
@@ -47,6 +49,12 @@ func resourcePostgreSQLExtension() *schema.Resource {
 				Computed:    true,
 				Description: "Sets the version number of the extension",
 			},
+			extDatabaseAttr: {
```
Comment (cyrilgdn) [inline postgresql/resource_postgresql_extension.go]: You need to set `ForceNew` to true. If we change the database we need to remove the extension from the old one and create it in the new one. (it's impossible to update)

**Comment by dvdliao** [comment]
done

**Comment by dvdliao** [comment]
thanks @cyrilgdn! I've committed the fix

**Comment by dvdliao** [comment]
hey @cyrilgdn do you know when this will be released? thanks

**Comment by cyrilgdn** [comment]
@dvdliao I've just asked for a new release so probably today.

**Comment by cyrilgdn** [comment]
@dvdliao FYI, version 1.1.0 [has been released](https://github.com/terraform-providers/terraform-provider-postgresql/releases/tag/v1.1.0)

**Pull request merged**